ElectronClient/gui/KeymapConfig/KeymapConfigScreen.tsx

```diff
@@ -0,0 +1,168 @@
+import * as React from 'react';
+import { useState, useEffect } from 'react';
+
+import styles_ from './styles';
+import { ShortcutRecorder } from './ShortcutRecorder';
+import CommandService from '../../lib/services/CommandService';
+import KeymapService, { KeymapItem } from '../../lib/services/KeymapService';
+
+const { _ } = require('lib/locale.js');
+const { shim } = require('lib/shim.js');
+
+const keymapService = KeymapService.instance();
+const commandService = CommandService.instance();
+
+interface CommandLabels {
+	[commandName: string]: string
+}
+
+// Some commands aren't actually registered in CommandService
+// These hardcoded labels are used for those commands instead
+const commandLabels: CommandLabels = {
+	quit: _('Quit'),
+	insertTemplate: _('Insert template'),
+	zoomActualSize: _('Actual Size'),
+	// Conditionally set the label because it's different in macOS and other platforms
+	config: _(shim.isMac() ? 'Preferences' : 'Options'),
+	gotoAnything: _('Goto Anything...'),
+	help: _('Website and documentation'),
+	hideApp: _('Hide Joplin'),
+	closeWindow: _('Close Window'),
+};
+// Obtain the rest of the labels
+keymapService.getCommands().forEach((commandName: string) => {
+	commandLabels[commandName] = commandLabels[commandName] || commandService.label(commandName);
+});
+
+interface CommandEditing {
+	[commandName: string]: boolean
+}
+
+export interface KeymapConfigScreenProps {
+	theme: number
+}
+
+export const KeymapConfigScreen = ({ theme }: KeymapConfigScreenProps) => {
+	const styles = styles_(theme);
+
+	const [filter, setFilter] = useState('');
+	const [errorMessage, setErrorMessage] = useState('');
+	const [keymap, setKeymap] = useState<KeymapItem[]>(() => keymapService.getKeymap());
+	const [editing, setEditing] = useState<CommandEditing>(() =>
+		keymapService.getCommands().reduce((accumulator: CommandEditing, command: string) => {
+			accumulator[command] = false;
+			return accumulator;
+		}, {})
+	);
+
+	const setCommandAccelerator = (commandName: string, accelerator: string) => {
+		setKeymap(prevKeymap => {
+			const newKeymap = [...prevKeymap];
+
+			newKeymap.find(item => item.command === commandName).accelerator = accelerator;
+			return newKeymap;
+		});
+	};
+
+	const resetCommandAccelerator = (commandName: string) => {
+		setKeymap(prevKeymap => {
+			const newKeymap = [...prevKeymap];
+			const defaultAccelerator = keymapService.getDefaultAccelerator(commandName);
+
+			newKeymap.find(item => item.command === commandName).accelerator = defaultAccelerator;
+			return newKeymap;
+		});
+	};
+
+	const toggleEditing = (commandName: string) => {
+		setEditing(prevEditing => {
+			const newEditing = { ...prevEditing };
+
+			newEditing[commandName] = !newEditing[commandName];
+			return newEditing;
+		});
+	};
```

Comment:
Regarding these three functions above, could you refactor using useReducer? I believe that’s really how it should be done in this case because a setter that calls a setter will be hard to maintain long term.

Reply:
Actually ignore this as once you’ll have refactored ShortcutRecorder you won’t need this layer of indirection, with setters calling setters, anymore.